File: src/message_parser.cpp
#include "message_parser.h"

#include <algorithm>

namespace MQTTTopics
{
MessageParser::Variables::Variables(const std::string& vehicleId, const std::string& deviceId, const std::string& transactionId) :
    vehicleId(vehicleId), deviceId(deviceId), transactionId(transactionId) {}

void MessageParser::TopicNode::addNode(const std::string& topic) {
    if(topic.empty()) {
        return;
    }

    auto slash = std::find(topic.begin(), topic.end(), '/');

    std::string subTopic = std::string(topic.begin(), slash);
    auto iter = this->adjacent.find(subTopic);

    if(iter == this->adjacent.end()) {
        iter = this->adjacent.emplace(subTopic, std::make_unique<TopicNode>(TopicNode())).first;
    }

    if(slash != topic.end()) {
        iter->second->addNode(std::string(slash + 1, topic.end()));
    }
}

std::unique_ptr<MessageParser::TopicNode>* MessageParser::TopicNode::findNode(const std::string& topic) {
    auto slash = std::find(topic.begin(), topic.end(), '/');

    std::string subTopic = std::string(topic.begin(), slash);
    auto iter = this->adjacent.find(subTopic);

    if(iter == this->adjacent.end()) {
        return nullptr;
    } 
    
    if(slash == topic.end()) {
        return &(iter->second);
    } else {
        return iter->second->findNode(std::string(slash + 1, topic.end()));
    }
}

std::vector<std::unique_ptr<MessageParser::TopicNode>*> MessageParser::TopicNode::findNodesVariables(const std::string& topic, const Variables& variables) {
    std::vector<std::unique_ptr<TopicNode>*> ret;
    this->findNodesVariablesRec(topic, variables, false, ret);

    return ret;
}

void MessageParser::TopicNode::findNodesVariablesRec(const std::string& topic, const Variables& variables, bool hashtag, std::vector<std::unique_ptr<TopicNode>*>& ret) {    
    auto slash = std::find(topic.begin(), topic.end(), '/');
    std::string subTopic = std::string(topic.begin(), slash);

    if(subTopic == "#") {
        hashtag = true;
    }

    if(hashtag) {
        for(auto& next : this->adjacent) {
            ret.push_back(&(next.second));
            next.second->findNodesVariablesRec("", variables, hashtag, ret);
        }
    } else {
        for(auto& next : this->adjacent) {
            bool match = false;

            if(subTopic == "+" || subTopic == next.first) {
                match = true;
            } else if(next.first == "<vehicleId>" && subTopic == variables.vehicleId) {
                match = true;
            } else if(next.first == "<deviceId>" && subTopic == variables.deviceId) {
                match = true;
            } else if(next.first == "<transactionId>" && subTopic == variables.transactionId) {
                match = true;
            }

            if(match) {
                if(slash == topic.end()) {
                    ret.push_back(&(next.second));
                } else {
                    next.second->findNodesVariablesRec(std::string(slash + 1, topic.end()), variables, hashtag, ret);
                }
            }
        }
    }
}

MessageParser::MessageParser() {
    this->tree.addNode("<vehicleId>");
    this->tree.addNode("<vehicleId>/<deviceId>");
    this->tree.addNode("<vehicleId>/<deviceId>/version");
    this->tree.addNode("<vehicleId>/<deviceId>/data");
    this->tree.addNode("<vehicleId>/<deviceId>/data/primary");
    this->tree.addNode("<vehicleId>/<deviceId>/data/secondary");
    this->tree.addNode("<vehicleId>/<deviceId>/data/bms");
    this->tree.addNode("<vehicleId>/<deviceId>/data/inverters");
    this->tree.addNode("<vehicleId>/<deviceId>/data/simulator");
    this->tree.addNode("<vehicleId>/<deviceId>/data/gps");
    this->tree.addNode("<vehicleId>/<deviceId>/data/brusa");
    this->tree.addNode("<vehicleId>/<deviceId>/data/temporary");
    this->tree.addNode("<vehicleId>/<deviceId>/data/cameraLogger");
    this->tree.addNode("<vehicleId>/<deviceId>/data/cameraLogger/can");
    this->tree.addNode("<vehicleId>/<deviceId>/data/cameraLogger/can/imuAngularRate");
    this->tree.addNode("<vehicleId>/<deviceId>/data/cameraLogger/can/imuAcceleration");
    this->tree.addNode("<vehicleId>/<deviceId>/data/cameraLogger/can/vehiclePosition");
    this->tree.addNode("<vehicleId>/<deviceId>/data/cameraLogger/can/vehicleSpeed");
    this->tree.addNode("<vehicleId>/<deviceId>/data/cameraLogger/can/frontAngularVelocity");
    this->tree.addNode("<vehicleId>/<deviceId>/data/cameraLogger/can/rearLeftAngularVelocity");
    this->tree.addNode("<vehicleId>/<deviceId>/data/cameraLogger/can/rearRightAngularVelocity");
    this->tree.addNode("<vehicleId>/<deviceId>/data/cameraLogger/can/steerAngle");
    this->tree.addNode("<vehicleId>/<deviceId>/data/cameraLogger/GPS");
    this->tree.addNode("<vehicleId>/<deviceId>/data/cameraLogger/GPS/hpposllh");
    this->tree.addNode("<vehicleId>/<deviceId>/data/cameraLogger/GPS/pvt");
    this->tree.addNode("<vehicleId>/<deviceId>/data/cameraLogger/GPS/relposned");
    this->tree.addNode("<vehicleId>/<deviceId>/extra_data_to_log");
    this->tree.addNode("<vehicleId>/<deviceId>/extra_tlm_data");
    this->tree.addNode("<vehicleId>/<deviceId>/extra_tlm_data/vehicleState");
    this->tree.addNode("<vehicleId>/<deviceId>/extra_tlm_data/baseline");
    this->tree.addNode("<vehicleId>/<deviceId>/extra_tlm_data/gpsMapOrigins");
    this->tree.addNode("<vehicleId>/<deviceId>/last_update");
    this->tree.addNode("<vehicleId>/<deviceId>/status");
    this->tree.addNode("<vehicleId>/<deviceId>/status/info");
    this->tree.addNode("<vehicleId>/<deviceId>/status/error");
    this->tree.addNode("<vehicleId>/<deviceId>/status/alert");
    this->tree.addNode("<vehicleId>/<deviceId>/status/canFrequencies");
    this->tree.addNode("<vehicleId>/<deviceId>/status/lapCounterStatus");
    this->tree.addNode("<vehicleId>/<deviceId>/status/lapCounterLaps");
    this->tree.addNode("<vehicleId>/<deviceId>/as");
    this->tree.addNode("<vehicleId>/<deviceId>/as/commands");
    this->tree.addNode("<vehicleId>/<deviceId>/as/commands/setValues");
    this->tree.addNode("<vehicleId>/<deviceId>/as/commands/setStatus");
    this->tree.addNode("<vehicleId>/<deviceId>/fileTransaction/request");
    this->tree.addNode("<vehicleId>/<deviceId>/fileTransaction/response");
    this->tree.addNode("<vehicleId>/<deviceId>/fileTransaction/<transactionId>");
    this->tree.addNode("<vehicleId>/<deviceId>/fileTransaction/<transactionId>/begin");
    this->tree.addNode("<vehicleId>/<deviceId>/fileTransaction/<transactionId>/end");
    this->tree.addNode("<vehicleId>/<deviceId>/fileTransaction/<transactionId>/chunk");
    this->tree.addNode("<vehicleId>/<deviceId>/fileTransaction/<transactionId>/chunk_ack");
    this->tree.addNode("<vehicleId>/<deviceId>/info");
    this->tree.addNode("<vehicleId>/<deviceId>/info/version");
    this->tree.addNode("<vehicleId>/<deviceId>/info/user");
    this->tree.addNode("<vehicleId>/<deviceId>/info/telemetryReplay");
    this->tree.addNode("<vehicleId>/<deviceId>/info/telemetryReplay/ready");
    this->tree.addNode("<vehicleId>/<deviceId>/info/telemetryReplay/start");
    this->tree.addNode("<vehicleId>/<deviceId>/info/session");
    this->tree.addNode("<vehicleId>/<deviceId>/info/session/started");
    this->tree.addNode("<vehicleId>/<deviceId>/info/session/stopped");
    this->tree.addNode("<vehicleId>/<deviceId>/info/session/logging");
    this->tree.addNode("<vehicleId>/<deviceId>/action");
    this->tree.addNode("<vehicleId>/<deviceId>/action/telemetryConfig");
    this->tree.addNode("<vehicleId>/<deviceId>/action/telemetryConfig/set");
    this->tree.addNode("<vehicleId>/<deviceId>/action/telemetryConfig/get");
    this->tree.addNode("<vehicleId>/<deviceId>/action/telemetryConfig/content");
    this->tree.addNode("<vehicleId>/<deviceId>/action/sessionConfig");
    this->tree.addNode("<vehicleId>/<deviceId>/action/sessionConfig/set");
    this->tree.addNode("<vehicleId>/<deviceId>/action/sessionConfig/get");
    this->tree.addNode("<vehicleId>/<deviceId>/action/sessionConfig/content");
    this->tree.addNode("<vehicleId>/<deviceId>/action/carConfig");
    this->tree.addNode("<vehicleId>/<deviceId>/action/carConfig/set");
    this->tree.addNode("<vehicleId>/<deviceId>/action/carConfig/get");
    this->tree.addNode("<vehicleId>/<deviceId>/action/carConfig/content");
    this->tree.addNode("<vehicleId>/<deviceId>/action/baselineConfig");
    this->tree.addNode("<vehicleId>/<deviceId>/action/baselineConfig/set");
    this->tree.addNode("<vehicleId>/<deviceId>/action/baselineConfig/get");
    this->tree.addNode("<vehicleId>/<deviceId>/action/baselineConfig/content");
    this->tree.addNode("<vehicleId>/<deviceId>/action/handcartSettings");
    this->tree.addNode("<vehicleId>/<deviceId>/action/handcartSettings/set");
    this->tree.addNode("<vehicleId>/<deviceId>/action/handcartSettings/get");
    this->tree.addNode("<vehicleId>/<deviceId>/action/handcartSettings/content");
    this->tree.addNode("<vehicleId>/<deviceId>/action/rawJsonConfig");
    this->tree.addNode("<vehicleId>/<deviceId>/action/rawJsonConfig/set");
    this->tree.addNode("<vehicleId>/<deviceId>/action/rawJsonConfig/get");
    this->tree.addNode("<vehicleId>/<deviceId>/action/rawJsonConfig/content");
    this->tree.addNode("<vehicleId>/<deviceId>/action/lapCounterConfig");
    this->tree.addNode("<vehicleId>/<deviceId>/action/lapCounterConfig/set");
    this->tree.addNode("<vehicleId>/<deviceId>/action/lapCounterConfig/get");
    this->tree.addNode("<vehicleId>/<deviceId>/action/lapCounterConfig/content");
    this->tree.addNode("<vehicleId>/<deviceId>/action/lapCounterTrack");
    this->tree.addNode("<vehicleId>/<deviceId>/action/lapCounterTrack/set");
    this->tree.addNode("<vehicleId>/<deviceId>/action/lapCounterTrack/get");
    this->tree.addNode("<vehicleId>/<deviceId>/action/lapCounterTrack/content");
    this->tree.addNode("<vehicleId>/<deviceId>/action/kill");
    this->tree.addNode("<vehicleId>/<deviceId>/action/start");
    this->tree.addNode("<vehicleId>/<deviceId>/action/reset");
    this->tree.addNode("<vehicleId>/<deviceId>/action/stop");
    this->tree.addNode("<vehicleId>/<deviceId>/action/startBaseline");
    this->tree.addNode("<vehicleId>/<deviceId>/action/stopBaseline");
    this->tree.addNode("<vehicleId>/<deviceId>/action/precharge");
    this->tree.addNode("<vehicleId>/<deviceId>/action/balance");
    this->tree.addNode("<vehicleId>/<deviceId>/action/stopBalance");
    this->tree.addNode("<vehicleId>/<deviceId>/action/charge");
    this->tree.addNode("<vehicleId>/<deviceId>/action/stopCharge");
    this->tree.addNode("<vehicleId>/<deviceId>/action/raw");
    this->tree.addNode("<vehicleId>/<deviceId>/action/resetLapcounter");
    this->tree.addNode("<vehicleId>/<deviceId>/action/setLapcounterStatus");
    this->tree.addNode("<vehicleId>/simulator");
    this->tree.addNode("<vehicleId>/simulator/inputs");
    this->tree.addNode("<vehicleId>/simulator/outputs");
    this->tree.addNode("<vehicleId>/simulator/initialState");
}

void MessageParser::setMessageParse(Topic topic, parse_t parse, void* argument) {
    auto node = this->tree.findNode(GetTopic(topic).topic);

    if(node != nullptr) {
        (*node)->parse = std::make_unique<parse_t>(parse);
        
        if(argument != nullptr) {
            (*node)->argument = argument; 
        }
    }
}

void MessageParser::parseMessage(const Variables& variables, const std::string& topic, const std::string& payload) {
    auto nodes = this->tree.findNodesVariables(topic, variables);

    for(auto& node : nodes) {
        if((*node)->parse != nullptr) {
            (*(*node)->parse)(payload, (*node)->argument);
        }
    }
}
}

File: inc/message_parser.py
from anytree import Node, RenderTree, LevelOrderIter, AsciiStyle, find


class MessageParser:

    def __init__(self):
        self.tree = None
        self.addNode("<vehicleId>",1)
        self.addNode("<vehicleId>/<deviceId>",2)
        self.addNode("<vehicleId>/<deviceId>/version",3)
        self.addNode("<vehicleId>/<deviceId>/data",3)
        self.addNode("<vehicleId>/<deviceId>/data/primary",4)
        self.addNode("<vehicleId>/<deviceId>/data/secondary",4)
        self.addNode("<vehicleId>/<deviceId>/data/bms",4)
        self.addNode("<vehicleId>/<deviceId>/data/inverters",4)
        self.addNode("<vehicleId>/<deviceId>/data/simulator",4)
        self.addNode("<vehicleId>/<deviceId>/data/gps",4)
        self.addNode("<vehicleId>/<deviceId>/data/brusa",4)
        self.addNode("<vehicleId>/<deviceId>/data/temporary",4)
        self.addNode("<vehicleId>/<deviceId>/data/cameraLogger",4)
        self.addNode("<vehicleId>/<deviceId>/data/cameraLogger/can",5)
        self.addNode("<vehicleId>/<deviceId>/data/cameraLogger/can/imuAngularRate",6)
        self.addNode("<vehicleId>/<deviceId>/data/cameraLogger/can/imuAcceleration",6)
        self.addNode("<vehicleId>/<deviceId>/data/cameraLogger/can/vehiclePosition",6)
        self.addNode("<vehicleId>/<deviceId>/data/cameraLogger/can/vehicleSpeed",6)
        self.addNode("<vehicleId>/<deviceId>/data/cameraLogger/can/frontAngularVelocity",6)
        self.addNode("<vehicleId>/<deviceId>/data/cameraLogger/can/rearLeftAngularVelocity",6)
        self.addNode("<vehicleId>/<deviceId>/data/cameraLogger/can/rearRightAngularVelocity",6)
        self.addNode("<vehicleId>/<deviceId>/data/cameraLogger/can/steerAngle",6)
        self.addNode("<vehicleId>/<deviceId>/data/cameraLogger/GPS",5)
        self.addNode("<vehicleId>/<deviceId>/data/cameraLogger/GPS/hpposllh",6)
        self.addNode("<vehicleId>/<deviceId>/data/cameraLogger/GPS/pvt",6)
        self.addNode("<vehicleId>/<deviceId>/data/cameraLogger/GPS/relposned",6)
        self.addNode("<vehicleId>/<deviceId>/extra_data_to_log",3)
        self.addNode("<vehicleId>/<deviceId>/extra_tlm_data",3)
        self.addNode("<vehicleId>/<deviceId>/extra_tlm_data/vehicleState",4)
        self.addNode("<vehicleId>/<deviceId>/extra_tlm_data/baseline",4)
        self.addNode("<vehicleId>/<deviceId>/extra_tlm_data/gpsMapOrigins",4)
        self.addNode("<vehicleId>/<deviceId>/last_update",3)
        self.addNode("<vehicleId>/<deviceId>/status",3)
        self.addNode("<vehicleId>/<deviceId>/status/info",4)
        self.addNode("<vehicleId>/<deviceId>/status/error",4)
        self.addNode("<vehicleId>/<deviceId>/status/alert",4)
        self.addNode("<vehicleId>/<deviceId>/status/canFrequencies",4)
        self.addNode("<vehicleId>/<deviceId>/status/lapCounterStatus",4)
        self.addNode("<vehicleId>/<deviceId>/status/lapCounterLaps",4)
        self.addNode("<vehicleId>/<deviceId>/as",3)
        self.addNode("<vehicleId>/<deviceId>/as/commands",4)
        self.addNode("<vehicleId>/<deviceId>/as/commands/setValues",5)
        self.addNode("<vehicleId>/<deviceId>/as/commands/setStatus",5)
        self.addNode("<vehicleId>/<deviceId>/fileTransaction/request",4)
        self.addNode("<vehicleId>/<deviceId>/fileTransaction/response",4)
        self.addNode("<vehicleId>/<deviceId>/fileTransaction/<transactionId>",4)
        self.addNode("<vehicleId>/<deviceId>/fileTransaction/<transactionId>/begin",5)
        self.addNode("<vehicleId>/<deviceId>/fileTransaction/<transactionId>/end",5)
        self.addNode("<vehicleId>/<deviceId>/fileTransaction/<transactionId>/chunk",5)
        self.addNode("<vehicleId>/<deviceId>/fileTransaction/<transactionId>/chunk_ack",5)
        self.addNode("<vehicleId>/<deviceId>/info",3)
        self.addNode("<vehicleId>/<deviceId>/info/version",4)
        self.addNode("<vehicleId>/<deviceId>/info/user",4)
        self.addNode("<vehicleId>/<deviceId>/info/telemetryReplay",4)
        self.addNode("<vehicleId>/<deviceId>/info/telemetryReplay/ready",5)
        self.addNode("<vehicleId>/<deviceId>/info/telemetryReplay/start",5)
        self.addNode("<vehicleId>/<deviceId>/info/session",4)
        self.addNode("<vehicleId>/<deviceId>/info/session/started",5)
        self.addNode("<vehicleId>/<deviceId>/info/session/stopped",5)
        self.addNode("<vehicleId>/<deviceId>/info/session/logging",5)
        self.addNode("<vehicleId>/<deviceId>/action",3)
        self.addNode("<vehicleId>/<deviceId>/action/telemetryConfig",4)
        self.addNode("<vehicleId>/<deviceId>/action/telemetryConfig/set",5)
        self.addNode("<vehicleId>/<deviceId>/action/telemetryConfig/get",5)
        self.addNode("<vehicleId>/<deviceId>/action/telemetryConfig/content",5)
        self.addNode("<vehicleId>/<deviceId>/action/sessionConfig",4)
        self.addNode("<vehicleId>/<deviceId>/action/sessionConfig/set",5)
        self.addNode("<vehicleId>/<deviceId>/action/sessionConfig/get",5)
        self.addNode("<vehicleId>/<deviceId>/action/sessionConfig/content",5)
        self.addNode("<vehicleId>/<deviceId>/action/carConfig",4)
        self.addNode("<vehicleId>/<deviceId>/action/carConfig/set",5)
        self.addNode("<vehicleId>/<deviceId>/action/carConfig/get",5)
        self.addNode("<vehicleId>/<deviceId>/action/carConfig/content",5)
        self.addNode("<vehicleId>/<deviceId>/action/baselineConfig",4)
        self.addNode("<vehicleId>/<deviceId>/action/baselineConfig/set",5)
        self.addNode("<vehicleId>/<deviceId>/action/baselineConfig/get",5)
        self.addNode("<vehicleId>/<deviceId>/action/baselineConfig/content",5)
        self.addNode("<vehicleId>/<deviceId>/action/handcartSettings",4)
        self.addNode("<vehicleId>/<deviceId>/action/handcartSettings/set",5)
        self.addNode("<vehicleId>/<deviceId>/action/handcartSettings/get",5)
        self.addNode("<vehicleId>/<deviceId>/action/handcartSettings/content",5)
        self.addNode("<vehicleId>/<deviceId>/action/rawJsonConfig",4)
        self.addNode("<vehicleId>/<deviceId>/action/rawJsonConfig/set",5)
        self.addNode("<vehicleId>/<deviceId>/action/rawJsonConfig/get",5)
        self.addNode("<vehicleId>/<deviceId>/action/rawJsonConfig/content",5)
        self.addNode("<vehicleId>/<deviceId>/action/lapCounterConfig",4)
        self.addNode("<vehicleId>/<deviceId>/action/lapCounterConfig/set",5)
        self.addNode("<vehicleId>/<deviceId>/action/lapCounterConfig/get",5)
        self.addNode("<vehicleId>/<deviceId>/action/lapCounterConfig/content",5)
        self.addNode("<vehicleId>/<deviceId>/action/lapCounterTrack",4)
        self.addNode("<vehicleId>/<deviceId>/action/lapCounterTrack/set",5)
        self.addNode("<vehicleId>/<deviceId>/action/lapCounterTrack/get",5)
        self.addNode("<vehicleId>/<deviceId>/action/lapCounterTrack/content",5)
        self.addNode("<vehicleId>/<deviceId>/action/kill",4)
        self.addNode("<vehicleId>/<deviceId>/action/start",4)
        self.addNode("<vehicleId>/<deviceId>/action/reset",4)
        self.addNode("<vehicleId>/<deviceId>/action/stop",4)
        self.addNode("<vehicleId>/<deviceId>/action/startBaseline",4)
        self.addNode("<vehicleId>/<deviceId>/action/stopBaseline",4)
        self.addNode("<vehicleId>/<deviceId>/action/precharge",4)
        self.addNode("<vehicleId>/<deviceId>/action/balance",4)
        self.addNode("<vehicleId>/<deviceId>/action/stopBalance",4)
        self.addNode("<vehicleId>/<deviceId>/action/charge",4)
        self.addNode("<vehicleId>/<deviceId>/action/stopCharge",4)
        self.addNode("<vehicleId>/<deviceId>/action/raw",4)
        self.addNode("<vehicleId>/<deviceId>/action/resetLapcounter",4)
        self.addNode("<vehicleId>/<deviceId>/action/setLapcounterStatus",4)
        self.addNode("<vehicleId>/simulator",2)
        self.addNode("<vehicleId>/simulator/inputs",3)
        self.addNode("<vehicleId>/simulator/outputs",3)
        self.addNode("<vehicleId>/simulator/initialState",3)

    def addNode(self, topic, layer):
        if topic != "":
            topic_split = topic.split("/")
            if self.tree != None:
                _parent = find(
                    self.tree,
                    lambda node: node.name == topic_split[len(topic_split) - 2]
                    and node.layer == layer - 1,
                )
                Node(
                    topic_split[len(topic_split) - 1],
                    parent=_parent,
                    arg=None,
                    layer=layer,
                    function=None,
                )
            else:
                self.tree = Node(
                    topic_split[len(topic_split) - 1],
                    arg=None,
                    layer=layer,
                    function=None,
                )

    def findNode(self, topic):
        if topic != "":
            topic_split = topic.split("/")
        topic_split[0] = "<vehicleId>"
        topic_split[1] = "<deviceId>"
        _topic_split = "/".join(topic_split)
        topic_split = _topic_split.split("/")
        _node = None
        match = [False] * len(topic_split)
        for i in range(len(topic_split)):
            for node in LevelOrderIter(self.tree, maxlevel=i + 1):
                if node.name == topic_split[i]:
                    _node = node
                    match[i] = True
        return _node

    def setMessageParse(self, topic, parse_function, argument):
        topic_split = topic.split("/")
        topic_split[0] = "<vehicleId>"
        topic_split[1] = "<deviceId>"
        _topic = "/".join(topic_split)
        _node = self.findNode(_topic)
        if _node != None:
            _node.function = parse_function
            _node.arg = argument

    def parseMessage(self, topic, payload):
        topic_split = topic.split("/")
        topic_split[0] = "<vehicleId>"
        topic_split[1] = "<deviceId>"
        _topic = "/".join(topic_split)
        _node = self.findNode(_topic)
        if _node != None and _node.function != None and _node.arg != None:
            _node.function(payload, _node.arg)